File: backend/app/schemas/privacy.py
"""Pydantic schemas for the GDPR / CCPA privacy endpoints (`/api/privacy`).

Request/response models for DSAR export, erasure / anonymization, and the
request-history list. The DSAR bundle is intentionally loosely typed (``dict``)
— it aggregates heterogeneous PII + related records across the control and
tenant DBs, and the privacy officer wants the raw portable JSON, not a flattened
view. The request-tracking responses are strict and PII-free.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.data_subject_request import (
    REQUEST_TYPES,
    SUBJECT_TYPES,
)

# Surfaced for the OpenAPI docs + validation error messages.
SUBJECT_TYPE_VALUES = list(SUBJECT_TYPES)
REQUEST_TYPE_VALUES = list(REQUEST_TYPES)


class DSARRequest(BaseModel):
    """Identify the data subject to export.

    ``identifier`` is an email for ``user`` / ``vendor_user`` subjects, or a
    Vendor UUID (string) for ``vendor_contact``. The resolver in
    ``services.privacy_export`` interprets it per ``subject_type``.
    """

    subject_type: str = Field(..., description=f"One of {SUBJECT_TYPE_VALUES}")
    identifier: str = Field(..., min_length=1, max_length=320)


class DSARResponse(BaseModel):
    """The portable DSAR bundle + the request-tracking metadata."""

    request_id: str
    subject_type: str
    subject_id: str
    generated_at: str
    # The portable bundle: every PII field + related-record summary held about
    # the subject, grouped by source. Loosely typed by design.
    data: dict


class ErasureRequest(BaseModel):
    """Identify the subject to erase + an explicit confirmation flag.

    ``confirm`` must be ``true`` — erasure is destructive (irreversible PII
    redaction), so the caller acknowledges intent. ``note`` is an optional,
    PII-free operator note (e.g. the legal basis / ticket reference).
    """

    subject_type: str = Field(..., description=f"One of {SUBJECT_TYPE_VALUES}")
    identifier: str = Field(..., min_length=1, max_length=320)
    confirm: bool = Field(..., description="Must be true to proceed")
    note: str | None = Field(default=None, max_length=500)


class ErasureResponse(BaseModel):
    request_id: str
    subject_type: str
    subject_id: str
    status: str  # completed | noop
    already_erased: bool
    fields_redacted: int
    # Non-identifying breakdown: which record kinds were touched + counts.
    record_counts: dict
    completed_at: str


class DataSubjectRequestSummary(BaseModel):
    """One row of the request-history list — strictly PII-free."""

    id: str
    request_type: str
    subject_type: str
    subject_id: str | None
    status: str
    requested_by: str | None
    fields_redacted: int
    record_counts: dict | None
    note: str | None
    created_at: datetime | None
    completed_at: datetime | None

    @classmethod
    def from_row(cls, row) -> DataSubjectRequestSummary:
        return cls(
            id=str(row.id),
            request_type=row.request_type,
            subject_type=row.subject_type,
            subject_id=str(row.subject_id) if row.subject_id else None,
            status=row.status,
            requested_by=str(row.requested_by) if row.requested_by else None,
            fields_redacted=row.fields_redacted,
            record_counts=row.record_counts,
            note=row.note,
            created_at=row.created_at,
            completed_at=row.completed_at,
        )


class DataSubjectRequestList(BaseModel):
    total: int
    requests: list[DataSubjectRequestSummary]

File: frontend/src/lib/types/privacy.ts
// Types for the GDPR/CCPA privacy surface (`/admin/privacy`). Mirrors
// `backend/app/schemas/privacy.py` + `backend/app/models/data_subject_request.py`.

export const SUBJECT_TYPES = ['user', 'vendor_user', 'vendor_contact'] as const;
export type SubjectType = (typeof SUBJECT_TYPES)[number];

export const SUBJECT_TYPE_LABELS: Record<SubjectType, string> = {
	user: 'AP-team user',
	vendor_user: 'Supplier-portal login',
	vendor_contact: 'Vendor contact details'
};

/** What the identifier field means for each subject type — shown as the
 *  field hint so an admin doesn't have to guess between an email and a UUID. */
export const SUBJECT_IDENTIFIER_HINTS: Record<SubjectType, string> = {
	user: "The user's email address.",
	vendor_user: "The supplier-portal login's email address.",
	vendor_contact: 'The Vendor UUID (from the vendor detail view or the URL).'
};

export interface DSARRequest {
	subject_type: SubjectType;
	identifier: string;
}

export interface DSARResponse {
	request_id: string;
	subject_type: string;
	subject_id: string;
	generated_at: string;
	// Loosely typed by design — a heterogeneous portable PII bundle.
	data: Record<string, unknown>;
}

export interface ErasureRequest {
	subject_type: SubjectType;
	identifier: string;
	confirm: true;
	note?: string;
}

export interface ErasureResponse {
	request_id: string;
	subject_type: string;
	subject_id: string;
	status: 'completed' | 'noop';
	already_erased: boolean;
	fields_redacted: number;
	record_counts: Record<string, unknown>;
	completed_at: string;
}

export type DataSubjectRequestType = 'dsar_export' | 'erasure';
export type DataSubjectRequestStatus = 'completed' | 'failed' | 'noop';

export interface DataSubjectRequestSummary {
	id: string;
	request_type: DataSubjectRequestType;
	subject_type: string;
	subject_id: string | null;
	status: DataSubjectRequestStatus;
	requested_by: string | null;
	fields_redacted: number;
	record_counts: Record<string, unknown> | null;
	note: string | null;
	created_at: string | null;
	completed_at: string | null;
}

export interface DataSubjectRequestList {
	total: number;
	requests: DataSubjectRequestSummary[];
}
